### PR TITLE
Fix a regression that caused us to listen to extra events at the top

### PR DESCRIPTION
**Pull request by gaearon**
I think this should fix the regression introduced in https://github.com/facebook/react/pull/12629 and that @sophiebits has patched over in https://github.com/facebook/react/pull/12877.

The first three commits are small refactors (I can remove them but I found the final form easier to read with these changes). The last one fixes the logic.

I'll need to work on some way to test this and verify it’s correct. Here’s the methodology with which I arrived at this solution:

1. Take a list of [all DOM event types in master](https://github.com/facebook/react/blob/master/packages/react-dom/src/events/DOMTopLevelEventTypes.js)

2. For each type, record whether it exists [in either of two mappings that were deleted in #12629](https://github.com/facebook/react/blob/1047980dca0830cd55e1622f3fbefc38aeaadb91/packages/react-dom/src/events/BrowserEventConstants.js): `topLevelTypes` or `mediaEventTypes`. Some types were new in master (pointer events) which I treat as if they were in `topLevelTypes`.

After doing this categorization, I ended up with four lists:

### NOT top AND NOT media

* `TOP_INVALID`
* `TOP_RESET`
* `TOP_SUBMIT`

### NOT top AND media

* `TOP_ABORT`
* `TOP_CAN_PLAY`
* `TOP_CAN_PLAY_THROUGH`
* `TOP_DURATION_CHANGE`
* `TOP_EMPTIED`
* `TOP_ENCRYPTED`
* `TOP_ENDED`
* `TOP_ERROR`
* `TOP_LOADED_DATA`
* `TOP_LOADED_METADATA`
* `TOP_PAUSE`
* `TOP_PLAY`
* `TOP_PLAYING`
* `TOP_PROGRESS`
* `TOP_RATE_CHANGE`
* `TOP_SEEKED`
* `TOP_SEEKING`
* `TOP_STALLED`
* `TOP_SUSPEND`
* `TOP_TIME_UPDATE`
* `TOP_VOLUME_CHANGE`
* `TOP_WAITING`

### top AND media

* `TOP_LOAD_START`

### top AND NOT media

* `TOP_ANIMATION_END`
* `TOP_ANIMATION_ITERATION`
* `TOP_ANIMATION_START`
* `TOP_BLUR`
* `TOP_CANCEL`
* `TOP_CHANGE`
* `TOP_CLICK`
* `TOP_CLOSE`
* `TOP_COMPOSITION_END`
* `TOP_COMPOSITION_START`
* `TOP_COMPOSITION_UPDATE`
* `TOP_CONTEXT_MENU`
* `TOP_COPY`
* `TOP_CUT`
* `TOP_DOUBLE_CLICK`
* `TOP_DRAG`
* `TOP_DRAG_END`
* `TOP_DRAG_ENTER`
* `TOP_DRAG_EXIT`
* `TOP_DRAG_LEAVE`
* `TOP_DRAG_OVER`
* `TOP_DRAG_START`
* `TOP_DROP`
* `TOP_FOCUS`
* `TOP_GOT_POINTER_CAPTURE`
* `TOP_INPUT`
* `TOP_KEY_DOWN`
* `TOP_KEY_PRESS`
* `TOP_KEY_UP`
* `TOP_LOAD`
* `TOP_LOST_POINTER_CAPTURE`
* `TOP_MOUSE_DOWN`
* `TOP_MOUSE_MOVE`
* `TOP_MOUSE_OUT`
* `TOP_MOUSE_OVER`
* `TOP_MOUSE_UP`
* `TOP_PASTE`
* `TOP_POINTER_CANCEL`
* `TOP_POINTER_DOWN`
* `TOP_POINTER_ENTER`
* `TOP_POINTER_LEAVE`
* `TOP_POINTER_MOVE`
* `TOP_POINTER_OUT`
* `TOP_POINTER_OVER`
* `TOP_POINTER_UP`
* `TOP_SCROLL`
* `TOP_SELECTION_CHANGE`
* `TOP_TEXT_INPUT`
* `TOP_TOGGLE`
* `TOP_TOUCH_CANCEL`
* `TOP_TOUCH_END`
* `TOP_TOUCH_MOVE`
* `TOP_TOUCH_START`
* `TOP_TRANSITION_END`
* `TOP_WHEEL`

From these lists, it follows that:

* “Media events should not be listened to on top level” is generally true.
* `TOP_LOAD_START` is an exception to the previous rule: even though it’s a media event, we listen to it on the top level.
* There are three additional events (`TOP_INVALID`, `TOP_RESET`, and `TOP_SUBMIT`) that are *neither* in the media nor the top whitelist. So we should exclude them *too*.

That’s reflected in my switch structure:

```js
      switch (dependency) {
        // ... other special cases ...
        case TOP_INVALID:
        case TOP_SUBMIT:
        case TOP_RESET:
          // We listen to them on the target DOM elements.
          // Some of them bubble so we don't want them to fire twice.
          break;
        case TOP_LOAD_START:
          // Even though it's a media event, it also exists on <img>.
          // So we need to listen at top level, unlike other media events.
          trapBubbledEvent(dependency, mountAt);
          break;
        default:
          // By default, listen on the top level to all non-media events.
          const isMediaEvent = mediaEventTypes.indexOf(dependency) !== -1;
          if (!isMediaEvent) {
            trapBubbledEvent(dependency, mountAt);
          }
          break;
      }
```

Note that I’m introducing an `indexOf` lookup there. I think it’s fine because this code only runs once per event, and `indexOf` is very fast on small arrays (`mediaEventTypes` has 23 elements). So I wouldn’t expect this to regress anything. But I can turn it into an object lookup if needed.

### CODE REVIEW

**Comment by react-linter** [comment]
ReactDOM: size: 0.0%, gzip: 0.0%


  <details>
  <summary>Details of bundled changes.</summary>

  <p>Comparing: ad27845ccd4fe34189514db23b645d7ab2213496...9363c2a7aa7571f7c11c775c612a47efc00c7b4d</p>


  
## react-dom
File | Filesize Diff | Gzip Diff | Prev Size | Current Size | Prev Gzip | Current Gzip | ENV
 ---  |  ---  |  ---  |  ---  |  ---  |  ---  |  ---  |  --- 
react-dom.development.js | +0.1% | +0.2% | 618.64 KB | 619.39 KB | 144.25 KB | 144.53 KB | UMD_DEV
**react-dom.production.min.js** | **0.0%** | **0.0%** | **93.79 KB** | **93.83 KB** | **30.36 KB** | **30.36 KB** | **UMD_PROD**
react-dom.development.js | +0.1% | +0.2% | 603.01 KB | 603.76 KB | 140.26 KB | 140.53 KB | NODE_DEV
**react-dom.production.min.js** | **0.0%** | **0.0%** | **92.29 KB** | **92.33 KB** | **29.34 KB** | **29.34 KB** | **NODE_PROD**
react-dom-test-utils.development.js | +0.5% | +0.8% | 45.04 KB | 45.27 KB | 12.37 KB | 12.47 KB | UMD_DEV
react-dom-test-utils.development.js | +0.6% | +0.9% | 39.9 KB | 40.13 KB | 10.94 KB | 11.04 KB | NODE_DEV
ReactDOM-dev.js | +0.1% | +0.2% | 624.79 KB | 625.54 KB | 142.89 KB | 143.17 KB | FB_WWW_DEV
**ReactDOM-prod.js** | **0.0%** | **-0.0%** | **274.02 KB** | **274.02 KB** | **51.6 KB** | **51.58 KB** | **FB_WWW_PROD**
ReactTestUtils-dev.js | +0.6% | +0.9% | 41.29 KB | 41.52 KB | 11.14 KB | 11.24 KB | FB_WWW_DEV

  </details>
  
<p align="right">
  Generated by :no_entry_sign: <a href="http://github.com/danger/danger-js/">dangerJS</a>
</p>

File: packages/react-dom/src/__tests__/ReactDOMEventListener-test.js [hunk 1]
```diff
@@ -217,4 +217,157 @@ describe('ReactDOMEventListener', () => {
...
+        // https://developer.mozilla.org/en-US/docs/Web/Events/reset
+        bubbles: true,
+      }),
+    );
+    expect(handleReset).toHaveBeenCalledTimes(1);
+
+    formRef.current.dispatchEvent(
+      new Event('submit', {
+        // https://developer.mozilla.org/en-US/docs/Web/Events/submit
+        bubbles: true,
```
Comment (sophiebits) [inline packages/react-dom/src/__tests__/ReactDOMEventListener-test.js]: Might be nice to have a bubbles-false test for this too if we don't.

File: packages/react-dom/src/__tests__/ReactDOMEventListener-test.js [hunk 2]
```diff
@@ -217,4 +217,157 @@ describe('ReactDOMEventListener', () => {
...
+      }),
+    );
+    expect(handleSubmit).toHaveBeenCalledTimes(1);
```
Comment (sophiebits) [inline packages/react-dom/src/__tests__/ReactDOMEventListener-test.js]: This failed before, right?

File: packages/react-dom/src/events/ReactBrowserEventEmitter.js [hunk 3]
```diff
@@ -130,29 +133,39 @@ export function listenTo(
...
+          }
+          break;
+        case TOP_INVALID:
+        case TOP_SUBMIT:
+        case TOP_RESET:
+          // We listen to them on the target DOM elements.
+          // Some of them bubble so we don't want them to fire twice.
+          break;
+        default:
+          // By default, listen on the top level to all non-media events.
```
Comment (sophiebits) [inline packages/react-dom/src/events/ReactBrowserEventEmitter.js]: Maybe add:

// Media events don't bubble so adding the listener wouldn't do anything.

File: packages/react-dom/src/events/ReactBrowserEventEmitter.js [hunk 4]
```diff
@@ -130,29 +133,39 @@ export function listenTo(
...
+          const isNonMediaEvent = mediaEventTypes.indexOf(dependency) === -1;
```
Comment (sophiebits) [inline packages/react-dom/src/events/ReactBrowserEventEmitter.js]: IMO this would be easier to read as

isMediaEvent
!== -1
if (!isMediaEvent)

Comment (gaearon) [reply packages/react-dom/src/events/ReactBrowserEventEmitter.js]: Force-pushed the version without the commit that changed this

File: packages/react-dom/src/__tests__/ReactDOMEventListener-test.js (hunk 2 above)
Comment (gaearon) [reply packages/react-dom/src/__tests__/ReactDOMEventListener-test.js]: Yes, it failed before your fix

File: packages/react-dom/src/__tests__/ReactDOMEventListener-test.js (hunk 1 above)
Comment (gaearon) [reply packages/react-dom/src/__tests__/ReactDOMEventListener-test.js]: When would it be false?

File: packages/react-dom/src/events/ReactBrowserEventEmitter.js (hunk 3 above)
Comment (gaearon) [reply packages/react-dom/src/events/ReactBrowserEventEmitter.js]: Added.

File: packages/react-dom/src/__tests__/ReactDOMEventListener-test.js (hunk 1 above)
Comment (sophiebits) [reply packages/react-dom/src/__tests__/ReactDOMEventListener-test.js]: Old browsers I think. That's why we listen on the element at all.

**Pull request merged**